abi/0.2.json

```diff
@@ -0,0 +1,313 @@
+        "args": {
+          "items": {
+            "$ref": "#/definitions/ArgumentDef"
+          },
+          "type": "array"
+        },
+        "kind": {
+          "enum": [
+            "Function"
+          ],
+          "type": "string"
+        },
+        "name": {
+          "type": "string"
+        },
+        "result": {
+          "$ref": "#/definitions/ResultDef"
+        }
+      },
+      "type": "object"
+    },
+    "ImportRefType": {
+      "properties": {
+        "import_id": {
+          "type": "string"
+        },
+        "kind": {
+          "enum": [
+            "ImportRef"
+          ],
+          "type": "string"
+        },
+        "ref_kind": {
+          "$ref": "#/definitions/UniqueDefKind"
+        },
+        "ref_name": {
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "MapType": {
+      "properties": {
+        "key": {
+          "$ref": "#/definitions/MapKeyScalarType"
+        },
+        "kind": {
+          "enum": [
+            "Map"
+          ],
+          "type": "string"
+        },
+        "value": {
+          "$ref": "#/definitions/OptionalType"
+        }
+      },
+      "type": "object"
+    },
+    "ObjectDef": {
+      "properties": {
+        "kind": {
+          "enum": [
+            "Object"
+          ],
+          "type": "string"
+        },
+        "name": {
+          "type": "string"
+        },
+        "props": {
+          "items": {
+            "$ref": "#/definitions/PropertyDef"
+          },
+          "type": "array"
+        }
+      },
+      "type": "object"
+    },
+    "OptionalType": {
+      "properties": {
+        "required": {
+          "type": "boolean"
+        },
+        "type": {
+          "$ref": "#/definitions/AnyType"
+        }
+      },
+      "type": "object"
+    },
+    "PropertyDef": {
+      "properties": {
+        "kind": {
+          "enum": [
+            "Property"
+          ],
+          "type": "string"
+        },
+        "name": {
+          "type": "string"
+        },
+        "required": {
+          "type": "boolean"
+        },
+        "type": {
+          "$ref": "#/definitions/AnyType"
+        }
+      },
+      "type": "object"
+    },
+    "RefType": {
+      "properties": {
+        "kind": {
+          "enum": [
+            "Ref"
+          ],
+          "type": "string"
+        },
+        "ref_kind": {
+          "$ref": "#/definitions/UniqueDefKind"
+        },
+        "ref_name": {
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "ResultDef": {
+      "properties": {
+        "kind": {
+          "enum": [
+            "Result"
+          ],
+          "type": "string"
+        },
+        "required": {
+          "type": "boolean"
+        },
+        "type": {
+          "$ref": "#/definitions/AnyType"
+        }
+      },
+      "type": "object"
+    },
+    "ScalarTypeEnum": {
+      "enum": [
+        "BigInt",
```

Comment:
BigInt, BigNumber, JSON, etc will be custom types and wouldn't be part of the default Scalar types

Reply:
Yeah I agree. But instead of rolling out one very big change, we split the work into smaller, easier to review stages.

This PR doesn't contain the changes to how we will handle custom types; that's being worked in another, more recent, branch and we want to merge it separately. This one only contains the changes to the general ABI structure along with other more general decisions, that you can find in the readme.